rules/DeadCode/SideEffect/SideEffectNodeDetector.php

```diff
@@ -48,40 +40,14 @@ public function __construct(
     ) {
     }
 
-    public function detect(Expr $expr, Scope $scope): bool
+    public function detect(Node $node, Scope $scope): bool
     {
-        if ($expr instanceof Assign) {
-            return true;
-        }
-
-        foreach (self::SIDE_EFFECT_NODE_TYPES as $sideEffectNodeType) {
-            if ($expr instanceof $sideEffectNodeType) {
-                return false;
-            }
-        }
-
-        $exprStaticType = $this->nodeTypeResolver->getType($expr);
-        if ($exprStaticType instanceof ConstantType) {
+        if (! $node instanceof Expr) {
             return false;
         }
 
-        if ($expr instanceof FuncCall) {
-            return ! $this->pureFunctionDetector->detect($expr, $scope);
-        }
-
-        if ($expr instanceof Variable || $expr instanceof ArrayDimFetch) {
-            $variable = $this->resolveVariable($expr);
-            // variables don't have side effects
-            return ! $variable instanceof Variable;
-        }
-
-        return true;
-    }
-
-    public function detectCallExpr(Node $node, Scope $scope): bool
-    {
-        if (! $node instanceof Expr) {
-            return false;
+        if ($node instanceof Assign) {
+            return $this->detect($node->expr, $scope);
```

Comment:
It can have side-effects for classes implementing magic-`__get()`

Reply:
That seems special use unrelated to this PR, and may already checked by `SIDE_EFFECT_NODE_TYPES` verify, but that will require more test case.

I am closing this for now to avoid regression, and collect more use case first based on previous PR:

- https://github.com/rectorphp/rector-src/pull/5373